solarwinds_apm/trace/base_metrics_processor.py

```diff
@@ -0,0 +1,101 @@
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+import logging
+from typing import TYPE_CHECKING
+
+from opentelemetry.sdk.trace import SpanProcessor
+from opentelemetry.semconv.trace import SpanAttributes
+from opentelemetry.trace import SpanKind, StatusCode
+
+from solarwinds_apm.apm_constants import INTL_SWO_SUPPORT_EMAIL
+from solarwinds_apm.w3c_transformer import W3CTransformer
+
+if TYPE_CHECKING:
+    from opentelemetry.sdk.trace import ReadableSpan
+
+    from solarwinds_apm.apm_txname_manager import SolarWindsTxnNameManager
+
+
+logger = logging.getLogger(__name__)
+
+
+class _SwBaseMetricsProcessor(SpanProcessor):
+    """Solarwinds base class for metrics span processors"""
+
+    _HTTP_METHOD = SpanAttributes.HTTP_METHOD  # "http.method"
+    _HTTP_STATUS_CODE = SpanAttributes.HTTP_STATUS_CODE  # "http.status_code"
+
+    _HTTP_SPAN_STATUS_UNAVAILABLE = 0
+    _TRANSACTION_NAME = "transaction_name"
+
+    def __init__(
+        self,
+        apm_txname_manager: "SolarWindsTxnNameManager",
+    ) -> None:
+        self.apm_txname_manager = apm_txname_manager
+
+    def get_trans_name_and_url_tran(
+        self,
+        span: "ReadableSpan",
+    ):
+        """Return cached trans_name and url_tran for current trace and span ID"""
+        txn_name_tuple = self.apm_txname_manager.get(
+            W3CTransformer.trace_and_span_id_from_context(span.context)
+        )
+        if not txn_name_tuple:
+            logger.error(
+                "Failed to retrieve transaction name for OTLP metrics generation. Please contact %s",
+                INTL_SWO_SUPPORT_EMAIL,
+            )
+            return None, None
+
+        try:
+            trans_name = txn_name_tuple[0]
+            url_tran = txn_name_tuple[1]
```

Comment:
I don't love how these are stored as tuples so I will change to a different data type in a separate PR.

Reply:
Addressed in https://github.com/solarwinds/apm-python/pull/255